File: lib/upload/upload-scenario.ts
// Upload sandbox — assembles a `DemoScenario`-shaped object for an
// arbitrary uploaded invoice, the same shape lib/queue/live-scenario.ts
// builds for the five seeded scenarios, so the existing result panels
// (ExtractedDataPanel, MatchEvidencePanel, ControlChecklist,
// ProposedActionPanel, AuditTrail) can render it unmodified — the plan's
// "point the workbench at the session invoice" instruction, made literal.
// Unlike getLiveScenario, there's no fixture to fall back to or overlay
// onto: every field here comes from the live DB row, because there's no
// other source of truth for a document nobody curated in advance.
import { supabaseAdmin } from "@/lib/supabase/server";
import {
  mapControl,
  mapAuditEvent,
  mapSupplier,
  mapExistingInvoice,
  fetchPurchaseOrder,
  fetchReceipt,
} from "@/lib/queue/live-scenario";
import type {
  AccountingChangeSet,
  DecisionOutcome,
  DemoScenario,
  DuplicateCandidate,
  ExtractedInvoice,
  InvoiceDecision,
  InvoiceMatchResult,
  MatchTier,
  PoMatchTier,
  SupplierMaster,
} from "@/lib/types";

type Db = ReturnType<typeof supabaseAdmin>;

// ledgerguard.md's security requirements: mask full bank account/routing
// numbers before they ever reach the browser. The seeded demo fixtures
// already only ever print "...ending 1234"-style values by construction,
// so this doesn't apply there — but an arbitrary uploaded invoice is real
// third-party content that may print full digit runs verbatim, in both the
// extracted value and its source quote.
function redactDigitRuns(text: string): string {
  return text.replace(/\d{5,}/g, (run) => "•".repeat(Math.max(0, run.length - 4)) + run.slice(-4));
}

function redactRemittanceField(field: ExtractedInvoice["remittanceDetails"]): ExtractedInvoice["remittanceDetails"] {
  if (!field) return field;
  return {
    ...field,
    value: field.value ? redactDigitRuns(field.value) : field.value,
    normalizedValue: field.normalizedValue ? redactDigitRuns(field.normalizedValue) : field.normalizedValue,
    evidence: field.evidence.map((e) => ({ ...e, text: redactDigitRuns(e.text) })),
  };
}

// Rendered when supplier tier is "none" — the common, honest case for an
// arbitrary uploaded document (ledgerguard.md's "honest product behavior").
// Every field reads as visibly absent rather than silently blank.
const UNMATCHED_SUPPLIER: SupplierMaster = {
  id: "unmatched",
  name: "Not matched to an approved supplier",
  taxId: "—",
  approvedDomain: "",
  status: "pending",
  bankOnFile: { bankName: "", accountLast4: "", routingLast4: "", verifiedAt: "" },
};

export type UploadScenarioStatus =
  | { state: "not_found" }
  | { state: "expired" }
  | { state: "processing" }
  | { state: "ready"; scenario: DemoScenario; expiresAt: string; originalFileName: string; storagePath: string | null };

/**
 * Loads one upload invoice by id, scoped to the owning session token so a
 * visitor can never fetch another visitor's result by guessing an id.
 * Returns a small state machine rather than throwing, since "not found",
 * "expired", and "still processing" are all expected, distinct UI states —
 * not error conditions.
 */
export async function getUploadScenario(invoiceId: string, sessionToken: string): Promise<UploadScenarioStatus> {
  const db: Db = supabaseAdmin();

  const { data: invoice } = await db
    .from("invoices")
    .select("*")
    .eq("id", invoiceId)
    .eq("source", "upload")
    .eq("session_token", sessionToken)
    .maybeSingle();

  if (!invoice) return { state: "not_found" };
  if (invoice.expires_at && new Date(invoice.expires_at) < new Date()) return { state: "expired" };

  const [{ data: controlRows }, { data: matchRow }, { data: decisionRow }, { data: auditRows }] = await Promise.all([
    db.from("controls").select("*").eq("invoice_id", invoice.id),
    db.from("match_results").select("*").eq("invoice_id", invoice.id).maybeSingle(),
    db.from("decisions").select("*").eq("invoice_id", invoice.id).maybeSingle(),
    db.from("audit_events").select("*").eq("invoice_id", invoice.id).order("timestamp"),
  ]);

  if (!decisionRow || !invoice.extracted || Object.keys(invoice.extracted as object).length === 0) {
    return { state: "processing" };
  }

  const supplierId = invoice.supplier_id ?? matchRow?.supplier_id ?? null;
  const purchaseOrderId = invoice.purchase_order_id ?? matchRow?.purchase_order_id ?? null;
  const firstReceiptId = (matchRow?.receipt_ids as string[] | null)?.[0] ?? null;
  const duplicateCandidates = (matchRow?.duplicate_candidates as DuplicateCandidate[] | null) ?? [];
  const firstDuplicateId = duplicateCandidates[0]?.existingInvoiceId ?? null;

  const [supplierRow, purchaseOrder, receipt, duplicateOfRow] = await Promise.all([
    supplierId ? db.from("suppliers").select("*").eq("id", supplierId).maybeSingle().then((r) => r.data) : Promise.resolve(null),
    purchaseOrderId ? fetchPurchaseOrder(db, purchaseOrderId) : Promise.resolve(undefined),
    firstReceiptId ? fetchReceipt(db, firstReceiptId) : Promise.resolve(undefined),
    firstDuplicateId ? db.from("invoices").select("*").eq("id", firstDuplicateId).maybeSingle().then((r) => r.data) : Promise.resolve(null),
  ]);

  const controls = (controlRows ?? []).map(mapControl);
  const outcome = decisionRow.outcome as DecisionOutcome;
  const rawExtracted = invoice.extracted as unknown as ExtractedInvoice;
  const extracted: ExtractedInvoice = {
    ...rawExtracted,
    remittanceDetails: redactRemittanceField(rawExtracted.remittanceDetails),
  };

  const match: InvoiceMatchResult = {
    supplierId: matchRow?.supplier_id ?? undefined,
    supplierMatch: (matchRow?.supplier_match as MatchTier | undefined) ?? "none",
    purchaseOrderId: matchRow?.purchase_order_id ?? undefined,
    purchaseOrderMatch: (matchRow?.purchase_order_match as PoMatchTier | undefined) ?? "none",
    receiptIds: (matchRow?.receipt_ids as string[] | null) ?? [],
    duplicateCandidates,
  };

  const decision: InvoiceDecision = {
    workflowId: decisionRow.workflow_id,
    outcome,
    reason: decisionRow.reason,
    controls,
    approvalRoute: (decisionRow.approval_route as string[] | null) ?? [],
    proposedAccountingChange: (decisionRow.proposed_accounting_change as AccountingChangeSet | null) ?? undefined,
    requiredActions: (decisionRow.required_actions as string[] | null) ?? [],
    policyVersion: decisionRow.policy_version,
  };

  const displayName = extracted.supplierName.value ?? invoice.original_file_name;
  const scenario: DemoScenario = {
    id: `upload:${invoice.id}`,
    order: 0,
    outcome,
    title: displayName,
    shortLabel: "Your upload",
    tagline: extracted.invoiceNumber.value ? `Invoice ${extracted.invoiceNumber.value}` : invoice.original_file_name,
    submission: {
      submissionId: invoice.submission_id,
      source: "upload",
      originalFileName: invoice.original_file_name,
      fileHash: invoice.file_hash,
      mimeType: "application/pdf",
      receivedAt: invoice.received_at,
    },
    documentLines: [], // real uploaded PDF is rendered directly, not from curated fixture lines
    extracted,
    supplier: supplierRow ? mapSupplier(supplierRow) : UNMATCHED_SUPPLIER,
    purchaseOrder,
    receipt,
    duplicateOf: duplicateOfRow ? mapExistingInvoice(duplicateOfRow) : undefined,
    match,
    controls,
    decision,
    auditEvents: (auditRows ?? []).map(mapAuditEvent),
    narrative: {
      whatHappened: decision.reason,
      whyItMatters:
        "This is your own document, run through the same real extraction, evidence-alignment, and control pipeline as the five seeded scenarios — under the upload sandbox's policy, which never auto-approves an uploaded document and treats an unrecognized supplier as an honest exception rather than a fabricated match.",
    },
  };

  return {
    state: "ready",
    scenario,
    expiresAt: invoice.expires_at ?? "",
    originalFileName: invoice.original_file_name,
    storagePath: invoice.storage_path,
  };
}

File: ledgerguard/upload/upload_scenario.py
"""Assembles a scenario-shaped object for an arbitrary uploaded invoice, the
same shape queue_.live_scenario builds for the five seeded scenarios, so
the existing result panels render it unmodified. Unlike get_live_scenario,
there's no fixture to fall back to or overlay onto: every field here comes
from the live DB row."""

import re
from datetime import UTC, datetime

from sqlalchemy import select
from sqlalchemy.orm import Session

from ..models import AuditEvent, Control, Decision, Invoice, MatchResult, Supplier, ensure_aware
from ..queue_.live_scenario import (
    fetch_purchase_order,
    fetch_receipt,
    lookup_duplicate_of,
    map_audit_event,
    map_control,
    map_existing_invoice,
    map_supplier,
)

# Masks full bank account/routing numbers before they ever reach the
# browser. The seeded demo fixtures only ever print "...ending 1234"-style
# values by construction, so this doesn't apply there — but an arbitrary
# uploaded invoice is real third-party content that may print full digit
# runs verbatim, in both the extracted value and its source quote.
_DIGIT_RUN = re.compile(r"\d{5,}")


def _redact_digit_runs(text: str) -> str:
    def repl(m: re.Match) -> str:
        run = m.group(0)
        return "•" * max(0, len(run) - 4) + run[-4:]

    return _DIGIT_RUN.sub(repl, text)


def _redact_remittance_field(field: dict | None) -> dict | None:
    if not field:
        return field
    redacted = dict(field)
    if redacted.get("value"):
        redacted["value"] = _redact_digit_runs(redacted["value"])
    if redacted.get("normalizedValue"):
        redacted["normalizedValue"] = _redact_digit_runs(redacted["normalizedValue"])
    redacted["evidence"] = [{**e, "text": _redact_digit_runs(e["text"])} for e in redacted.get("evidence", [])]
    return redacted


# Rendered when supplier tier is "none" — the common, honest case for an
# arbitrary uploaded document. Every field reads as visibly absent rather
# than silently blank.
UNMATCHED_SUPPLIER = {
    "id": "unmatched",
    "name": "Not matched to an approved supplier",
    "taxId": "—",
    "approvedDomain": "",
    "status": "pending",
    "bankOnFile": {"bankName": "", "accountLast4": "", "routingLast4": "", "verifiedAt": ""},
}


def get_upload_scenario(db: Session, invoice_id, session_token: str) -> dict:
    """Loads one upload invoice by id, scoped to the owning session token so
    a visitor can never fetch another visitor's result by guessing an id.
    Returns {"state": "not_found" | "expired" | "processing" | "ready", ...}
    — "not found", "expired", "still processing" are all expected UI
    states, not error conditions."""
    invoice = db.scalar(
        select(Invoice).where(Invoice.id == invoice_id, Invoice.source == "upload", Invoice.session_token == session_token)
    )

    if not invoice:
        return {"state": "not_found"}
    if invoice.expires_at and ensure_aware(invoice.expires_at) < datetime.now(UTC):
        return {"state": "expired"}

    control_rows = db.scalars(select(Control).where(Control.invoice_id == invoice.id)).all()
    match_row = db.scalar(select(MatchResult).where(MatchResult.invoice_id == invoice.id))
    decision_row = db.scalar(select(Decision).where(Decision.invoice_id == invoice.id))
    audit_rows = db.scalars(select(AuditEvent).where(AuditEvent.invoice_id == invoice.id).order_by(AuditEvent.created_at)).all()

    if not decision_row or not invoice.extracted:
        return {"state": "processing"}

    supplier_id = invoice.supplier_id or (match_row.supplier_id if match_row else None)
    purchase_order_id = invoice.purchase_order_id or (match_row.purchase_order_id if match_row else None)
    first_receipt_id = (match_row.receipt_ids or [None])[0] if match_row else None
    duplicate_candidates = match_row.duplicate_candidates or [] if match_row else []
    first_duplicate_id = duplicate_candidates[0]["existingInvoiceId"] if duplicate_candidates else None

    supplier_row = db.get(Supplier, supplier_id) if supplier_id else None
    purchase_order = fetch_purchase_order(db, purchase_order_id) if purchase_order_id else None
    receipt = fetch_receipt(db, first_receipt_id) if first_receipt_id else None
    duplicate_of_row = lookup_duplicate_of(db, first_duplicate_id) if first_duplicate_id else None

    controls = [map_control(c) for c in control_rows]
    outcome = decision_row.outcome
    extracted = dict(invoice.extracted)
    extracted["remittanceDetails"] = _redact_remittance_field(extracted.get("remittanceDetails"))

    match = {
        "supplierId": match_row.supplier_id if match_row else None,
        "supplierMatch": (match_row.supplier_match if match_row else None) or "none",
        "purchaseOrderId": match_row.purchase_order_id if match_row else None,
        "purchaseOrderMatch": (match_row.purchase_order_match if match_row else None) or "none",
        "receiptIds": match_row.receipt_ids or [] if match_row else [],
        "duplicateCandidates": duplicate_candidates,
    }

    decision = {
        "workflowId": decision_row.workflow_id,
        "outcome": outcome,
        "reason": decision_row.reason,
        "controls": controls,
        "approvalRoute": decision_row.approval_route or [],
        "proposedAccountingChange": decision_row.proposed_accounting_change,
        "requiredActions": decision_row.required_actions or [],
        "policyVersion": decision_row.policy_version,
    }

    display_name = (extracted.get("supplierName") or {}).get("value") or invoice.original_file_name
    invoice_number_value = (extracted.get("invoiceNumber") or {}).get("value")

    scenario = {
        "id": f"upload:{invoice.id}",
        "order": 0,
        "outcome": outcome,
        "title": display_name,
        "shortLabel": "Your upload",
        "tagline": f"Invoice {invoice_number_value}" if invoice_number_value else invoice.original_file_name,
        "submission": {
            "submissionId": invoice.submission_id,
            "source": "upload",
            "originalFileName": invoice.original_file_name,
            "fileHash": invoice.file_hash,
            "mimeType": "application/pdf",
            "receivedAt": invoice.received_at.isoformat(),
        },
        "documentLines": [],  # the real uploaded PDF is rendered directly, not from curated fixture lines
        "extracted": extracted,
        "supplier": map_supplier(supplier_row) if supplier_row else UNMATCHED_SUPPLIER,
        "purchaseOrder": purchase_order,
        "receipt": receipt,
        "duplicateOf": map_existing_invoice(duplicate_of_row) if duplicate_of_row else None,
        "match": match,
        "controls": controls,
        "decision": decision,
        "auditEvents": [map_audit_event(a) for a in audit_rows],
        "narrative": {
            "whatHappened": decision["reason"],
            "whyItMatters": (
                "This is your own document, run through the same real extraction, evidence-alignment, and "
                "control pipeline as the five seeded scenarios — under the upload sandbox's policy, which "
                "never auto-approves an uploaded document and treats an unrecognized supplier as an honest "
                "exception rather than a fabricated match."
            ),
        },
    }

    return {
        "state": "ready",
        "scenario": scenario,
        "expiresAt": invoice.expires_at.isoformat() if invoice.expires_at else "",
        "originalFileName": invoice.original_file_name,
        "storagePath": invoice.storage_path,
    }
